lib/assert.js

```diff
@@ -931,6 +933,86 @@ assert.doesNotMatch = function doesNotMatch(string, regexp, message) {
   internalMatch(string, regexp, message, doesNotMatch);
 };
 
+const noop = () => {};
+
+class CallTracker {
+
+  constructor() {
+    this.callChecks = new SafeSet();
+  }
+
+  calls(fn, exact = 1) {
+    if (process._exiting)
+      throw new ERR_UNAVAILABLE_DURING_EXIT();
+    if (typeof fn === 'number') {
+      exact = fn;
+      fn = noop;
+    } else if (fn === undefined) {
+      fn = noop;
+    }
+
+    validateUint32(exact, 'exact', true);
+
+    const context = {
+      exact,
+      actual: 0,
+      // eslint-disable-next-line no-restricted-syntax
+      stackTrace: new Error(),
+      name: fn.name || 'calls'
+    };
+    this.callChecks.add(context);
+    const callChecks = this.callChecks;
+
+    return function() {
+      // Increment call count
+      context.actual++;
+      if (context.actual === context.exact) {
+        // Once function has reached its call count remove it from
+        // callChecks set to prevent memory leaks
+        callChecks.delete(context);
+      }
+      // If function has been called more than expected times, add back into
+      // callchecks.
+      if (context.actual === context.exact + 1) {
+        callChecks.add(context);
+      }
+      return fn.apply(this, arguments);
+    };
+  }
+
+  report() {
+    // Create array for errors
+    const errors = [];
+    for (const context of this.callChecks) {
+      // If functions have not been called exact times
+      if (context.actual !== context.exact) {
+        const message = `Expected the ${context.name} function to be ` +
+        `executed ${context.exact} time(s) but was executed ` +
+        `${context.actual} time(s).`;
+        errors.push(new AssertionError({
+          message,
+          actual: context.actual,
+          expected: context.exact,
+          operator: context.name,
+          startStackFn: context.stackTrace
+        }));
+      }
+    }
+    if (errors.length > 0)
+      return errors;
+    else
+      return 0;
+  }
+
+  verify() {
+    const errors = this.report();
+    if (errors !== 0)
+      throw errors;
```

Comment:
Is this what `mustCall()` do? Throw an array? I think throwing a non-Error is a bad practice.

Reply:
I'm not familiar with any way to throw multiple errors at the same time. I tried compiling an array of the details for each function that wasn't called the expected number of times and then passing that to `throw new AssertionError()` but that just throws a single error with `undefined` values for everything. Have you any suggestions?

Reply:
@ConorDavenport ... for this what we should do is create a single error with a `details` property that is set to the array of individual errors returned by report. In other words... something like..

```js
  if (errors.length > 0) {
    const err = // create new assertion error
    err.details = errors;
    throw err;
  }
```

This way there is only a single error thrown but the details can be inspected.